src/param.rs

```diff
@@ -226,19 +262,88 @@ impl AudioParamProcessor {
+        //  However it is more convenient to put it there testing reasons.
+        //  review AudioParam mock in tests and review test_exponential_ramp_to_zero
+        //  accordingly.
+        if event.event_type == AutomationType::ExponentialRampToValueAtTime && event.value == 0. {
+            panic!(
+                "RangeError: Failed to execute 'exponentialRampToValueAtTime'
+                on 'AudioParam': The float target value provided (0) should not be
+                in the range ({:+e}, {:+e})",
+                -f32::MIN_POSITIVE,
+                f32::MIN_POSITIVE
+            );
+        }
+
+        // if no event in the timeline and event_type is `LinearRampToValueAtTime`
+        // or `ExponentialRampToValue` at time, we must insert a `SetValueAtTime`
+        // with intrisic value and calling time.
+        // cf. https://www.w3.org/TR/webaudio/#dom-audioparam-linearramptovalueattime
+        // cf. https://www.w3.org/TR/webaudio/#dom-audioparam-exponentialramptovalueattime
+        //
+        // @note - time=0 is what blink does but this is not compliant to the
+        // spec: time should be callTime. However this would imply a strong
+        // relation with `audio_context.curret_time` which breaks the tests
+        // -> probably ok for now as it only concerns edge cases (e.g. triggering
+        // a ramp in the future without an explicit `setTarget` before).
+        // -> at least we make sure that we have a previous event defined
+        //
+        // maybe this could also handled at last minute by checking if last_event exists, e.g.:
+        // ```
+        // if !last_event {
+        //     let automationEvent = events.pop();
+        //     let setTargetEvent { time: ts };
+        //     event.push(setTargetEvent);
+        //     event.push(automationEvent);
+        // }
+        // ```
+        if self.events.is_empty()
+            && (event.event_type == AutomationType::LinearRampToValueAtTime
+                || event.event_type == AutomationType::ExponentialRampToValueAtTime)
+        {
+            let set_value_event = AutomationEvent {
+                event_type: AutomationType::SetValueAtTime,
+                value: self.value,
+                time: 0., // make sure the event is applied before any other event
+            };
+
+            self.events.push(set_value_event);
+        }
+
+        // @todo - many checks need to be done for setValueCurveAtTime events
+
+        self.events.push(event);
+    }
+
     fn tick(&mut self, ts: f64, dt: f64, count: usize) -> &[f32] {
-        // store incoming automation events in sorted queue
-        for event in self.receiver.try_iter() {
-            self.events.push(event);
+        // println!("> tick - ts: {}, dt: {}, count: {}", ts, dt, count);
+        // handle incoming automation events in sorted queue
+        let events: Vec<AutomationEvent> = self.receiver.try_iter().collect();
```

Comment:
Ideally we avoid the allocation (`collect()`)

Reply:
ok thanks I didn't know that (I'm still not really confident with how rust works with memory I must confess)

Reply:
The reason we try to avoid allocations in the render thread is because the allocator has unreliable performance. An allocator is shared by all threads (typically in an `Arc<Mutex<Allocator>>`) so the audio thread competes with all other threads to get a lock on the allocator. In a CPU-heavy game, for example, this may cause a hiccup in the audio stream.

Interacting with the allocator comes in 3 flavours:

**Allocating:** any function that returns a heap allocated container (Vec, HashMap, Box, Rc etc see https://doc.rust-lang.org/alloc/index.html), for example `Vec/HashMap::new, collect(), split_off(), clone(), to_vec()`

**Deallocating:** any time a heap allocated object goes out of scope

**Reallocating:** If you're adding items to a container (Vec, Map, Set, Heap) and the current max capacity is reached. `push() insert() extend_from_slice() etc`

In the render thread it should be doable to avoid allocating and deallocating. For example see https://github.com/orottier/web-audio-api-rs/pull/17/files

Avoiding reallocations is tough. A strategy would be to pre-allocate all containers before launching the render thread but that would impose a maximum number of audio nodes and connections on the user.

Reply:
Thanks for the informations! It makes sens

So if I understand well, there might be a problem there:  https://github.com/orottier/web-audio-api-rs/blob/8f8f52ac5ceeb443303c905ca8e51e45a1125ded/src/param.rs#L140         

Because even if `ChannelData` has a "has copy-on-write semantics," as `output` is modified, new memory will be allocated, no? In that case, there should be a proper `ChannelData` allocated once for all for the param `output`.

Or am I missing something?

Reply:
Just for information, I just tried this (which looks more like an in-place write of `output[0]` to me):

> in `AudioParamProcessor.process()`
```rs
        let output = &mut outputs[0];
        output
            .channel_data_mut(0)
            .copy_from_slice(param_intrisic_value);
        output.add(input, ChannelInterpretation::Discrete);
```

instead of:

```rs
        let mut param_computed_value = inputs[0].clone(); // get new buf
        param_computed_value.force_mono();
        param_computed_value
            .channel_data_mut(0)
            .copy_from_slice(param_intrisic_value);
        // mix incoming signal
        param_computed_value.add(input, ChannelInterpretation::Discrete);

        outputs[0] = param_computed_value;
```

all test are passing and `many_oscillator_with_env` still sounds ok, but I don't if I'm missing something and how to benchmark that tough

Reply:
Ah, I forgot to mention, ChannelData is really special.

Have a good look at alloc.rs. If a ChannelData goes out of scope, it is stored in a pool of readily available buffers. Allocating (or invoking copy on write) a new ChannelData is very cheap if the pool is filled. The pool is used exclusively in the render thread so there are no locks involved.
Sorry, should have mentioned that!

(and now thinking, I should probably add an architecture.md file to this repo...)

Reply:
Your change in `AudioParamProcessor.process()` looks good though, much more readable!

Reply:
Thanks, I didn't understood the pool pattern indeed, good to know (`alloc.rs` slowly starts to make sens...)